### PR TITLE
Coinview prefetching

### PR DESCRIPTION
**Pull request by noescape00**
added prefetching (CoinviewPrefetchManager.cs) 

Fix: #1963

### CODE REVIEW

**Comment by dangershony** [comment]
Great you found a good solution to do it at high level, will review better tomorrow.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelprer.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Features.Consensus.CoinViews
+{
+    public class CoinviewHelprer
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelprer.cs]: `Helper`

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        /// </summary>
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
+        {
+            ChainedHeader currentHeader = header;
+
+            // Go Lookahead blocks ahead of current header and get block for prefetching.
+            // There might be several blocks at height of header.Height + Lookahead but
+            // only first one will be prefetched since prefetching is in place mostly to
+            // speed up IBD in which we can have only one chain on Stratis since we have
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: I would end the sentence after `speed up IBD` and start a new one, so it reads better. The comment makes sense but it's wordy.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+namespace Stratis.Bitcoin.Features.Consensus
+{
+    /// <summary>Fetches coins for blocks that are likely to be validated in the near future to speedup full validation.</summary>
+    public class CoinviewPrefetchManager : IDisposable
+    {
+        /// <summary>
+        /// How many blocks ahead prefetching will look.
+        /// When header at height X is dequeued block with header at height <c>X + Lookahead</c> will be prefetched in case block data is downloaded.
+        /// </summary>
+        private const int Lookahead = 5;
```
[the remark this answers is not in the document]
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: Actually, after evaluating the rest of the code - we are just looking at a single block a fixed height above the one whose header has currently been queued for prefetching?

**Comment by zeptin** [comment]
I have reviewed this and it looks good with some minor nits, but there should be some other eyes on it before approval perhaps.

I tried out the prefetch function without running all the tests, it seems to work as expected.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus
+{
+    /// <summary>Fetches coins for blocks that are likely to be validated in the near future to speedup full validation.</summary>
+    public class CoinviewPrefetchManager : IDisposable
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: I prefer the name to be `CoinviewPrefetcher` manager is used for in another context

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
+        {
+            ChainedHeader currentHeader = header;
+
+            // Go Lookahead blocks ahead of current header and get block for prefetching.
+            // There might be several blocks at height of header.Height + Lookahead but
+            // only first one will be prefetched since prefetching is in place mostly to
+            // speed up IBD in which we can have only one chain on Stratis since we have
+            // checkpoints and several chains on BTC but it's a rare case because creating
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: No need to mention stratis of bitcoin, also mostly there are not going to be two chains so its a very minor detail

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: >Actually, after evaluating the rest of the code - we are just looking at a single block a fixed height above the one whose header has currently been queued for prefetching?

Yes, correct.

[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: The value is random. I'm currently measuring performance using different values to check which gives the best result.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs (hunk 2 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelprer.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelprer.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs (hunk 4 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs (hunk 5 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/Rules/PowConsensusRuleEngine.cs [hunk 6]
```diff
@@ -57,8 +60,24 @@ public override Task Initialize()
             return ((DBreezeCoinView)((CachedCoinView)this.UtxoSet).Inner).InitializeAsync();
         }
 
+        public override async Task<ValidationContext> FullValidationAsync(ChainedHeader header, Block block)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/Rules/PowConsensusRuleEngine.cs]: brilliant

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            if (block == null)
+            {
+                this.logger.LogTrace("(-)[NO_BLOCK_DATA]");
+                return;
+            }
+
+            bool enforceBIP30 = DeploymentFlags.EnforceBIP30ForBlock(currentHeader);
+            uint256[] idsToFetch = this.coinviewHelper.GetIdsToFetch(block, enforceBIP30);
+
+            if (idsToFetch.Length != 0 && !cancellation.IsCancellationRequested)
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            ChainedHeader currentHeader = header;
+
+            // Go Lookahead blocks ahead of current header and get block for prefetching.
+            // There might be several blocks at height of header.Height + Lookahead but
+            // only first one will be prefetched since prefetching is in place mostly to
+            // speed up IBD in which we can have only one chain on Stratis since we have
+            // checkpoints and several chains on BTC but it's a rare case because creating
+            // alternative chains requires PoW which is expensive.
+            for (int i = 0; i < Lookahead; i++)
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: We are not prefetching 5 blocks, we are prefetching 1 block which is at height of current header + 5.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            {
+                if (currentHeader.Next.Count == 0)
+                {
+                    this.logger.LogTrace("(-)[NO_HEADERS]");
+                    return;
+                }
+
+                currentHeader = currentHeader.Next[0];
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: good catch. fixed

File: src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs [hunk 10]
```diff
@@ -191,7 +191,7 @@ public ValidationContext IntegrityValidation(ChainedHeader header, Block block)
         }
 
         /// <inheritdoc/>
-        public async Task<ValidationContext> FullValidationAsync(ChainedHeader header, Block block)
+        public virtual async Task<ValidationContext> FullValidationAsync(ChainedHeader header, Block block)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs]: Can you also add virtual to the other validation methods?

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        /// </summary>
+        /// <param name="header">Header of a block that was fully validated and requires prefetching.</param>
+        public void Prefetch(ChainedHeader header)
+        {
+            this.headersQueue.Enqueue(header);
+        }
+
+        public void Dispose()
+        {
+            this.headersQueue.Dispose();
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: it won't complete. Single item will be processed (if it's currently being processed already) and that's it.

File: src/Stratis.Bitcoin.Features.Consensus/Rules/PowConsensusRuleEngine.cs (hunk 6 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/Rules/PowConsensusRuleEngine.cs]: 🙂

File: src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs (hunk 10 above)
Comment (noescape00) [reply src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs]: fixed

**Comment by noescape00** [comment]
Don't merge yet. Still measuring performance.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            if (block == null)
+            {
+                this.logger.LogTrace("(-)[NO_BLOCK_DATA]");
+                return;
+            }
+
+            bool enforceBIP30 = DeploymentFlags.EnforceBIP30ForBlock(currentHeader);
+            uint256[] idsToFetch = this.coinviewHelper.GetIdsToFetch(block, enforceBIP30);
+
+            bool farFromTip = currentHeader.Height > this.chain.Tip.Height + 2;
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Why is this necessary?
currentHeader is always tip +5

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetchManager.cs]: ah yes sorry I read it wrong, yes thats good.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 12 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Because we are working in background thread and we might advance consensus too fast so by the time we want to prefetch it's already not needed.

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: hmm good point, its probably not worth adding a dependency on CH for that though.
Can you use chain instead?

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Didn't quite understand what you mean

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Don't use `ConcurrentChain`

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus
+{
+    /// <summary>Fetches coins for blocks that are likely to be validated in the near future to speedup full validation.</summary>
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: speed-up

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    public class CoinviewPrefetcher : IDisposable
+    {
+        /// <summary>
+        /// How many blocks ahead prefetching will look.
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: pre-fetching

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+            this.headersQueue = new AsyncQueue<ChainedHeader>(this.OnHeaderEnqueuedAsync);
+            this.coinviewHelper = new CoinviewHelper();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Prefetches UTXOs for a block and some blocks with higher height.
+        /// Prefetching is done on the background.
+        /// </summary>
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Pre-fetching (two line above).

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: pre-fetching

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
+        {
+            ChainedHeader currentHeader = header;
+
+            // Go Lookahead blocks ahead of current header and get block for prefetching.
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Go ?

Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Finding it hard to understand this line.

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+            this.coinviewHelper = new CoinviewHelper();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Prefetches UTXOs for a block and some blocks with higher height.
+        /// Prefetching is done on the background.
+        /// </summary>
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Log header details - or will this be redundant when Fody interceptors are in place ?

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                this.logger.LogTrace("(-)[NO_BLOCK_DATA]");
+                return;
+            }
+
+            bool farFromTip = currentHeader.Height > this.chain.Tip.Height + Lookahead/2;
+
+            if (!farFromTip)
+            {
+                this.logger.LogDebug("Far from tip is false! Skipping prefetching.");
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: Maybe "The current header is far from the tip - by N blocks! Skipping pre-fetching".

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+            var ids = new HashSet<uint256>();
+            foreach (Transaction tx in block.Transactions)
+            {
+                if (enforceBIP30)
+                {
+                    uint256 txId = tx.GetHash();
+                    ids.Add(txId);
+                }
+
+                if (!tx.IsCoinBase)
```
Comment (fenix2222) [inline src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs]: can improve nesting to make it a bit more flat

``` 
if (tx.IsCoinbase) continue;
foreach (TxIn input in tx.Inputs)
    ids.Add(input.PrevOut.Hash);
```

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+            Block block = currentHeader.Block;
+
+            if (block == null)
+            {
+                this.logger.LogTrace("(-)[NO_BLOCK_DATA]");
+                return;
+            }
+
+            bool farFromTip = currentHeader.Height > this.chain.Tip.Height + Lookahead/2;
```
Comment (fenix2222) [inline src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: don't we need brackets and spaces for `Lookahead/2`?

**Comment by noescape00** [comment]
Measurements: 

Master branch-- stratis mainnet sync time (500k): 22 m
Prefetch branch- stratis mainnet sync time (500k): 21m

Master BTC mainnet (250k): 2h 40m
Prefetch BTC mainnet (250k):  2h 24

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 13 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 14 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 15 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 16 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 17 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: We are at X and X+LookAhead is Y. So go from X to Y and prefetch it

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 18 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: >Log header details - or will this be redundant when Fody interceptors are in place ?

Yes. That's why there are no entry\exit logs

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 19 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: It's about not being far.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs (hunk 20 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs]: This code was moved. Tried to avoid any changes

File: src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs (hunk 21 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs]: fixed

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs (hunk 20 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs]: why does everybody hates nesting :D?

**Comment by noescape00** [comment]
It's ready for merge.

File: src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs (hunk 20 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/CoinViews/CoinviewHelper.cs]: lol we should focus on readability not optimising how we write code, the compiler anyway does lots of optimisation for us.
